### ISSUE TITLE
Nightly builds???

### ISSUE THREAD

**Issue by SANSYBOY1234** [pencil2d/pencil#582]
so the nightly builds arnt really nightly their like monthly builds.... so i was just asking if you guys could actually make them NIGHTLY

**Comment by Jose-Moreno** [comment]
@SANSYBOY1234 You have to realize this is basically a volunteer project, not a commercial product. There are limited human resources and basically only the Lead developer is coding and creating these builds for both Windows and MAC when he is able to. This is a time consuming process, so it's not like he can churn out daily builds after changing three lines of code. 

However should you or anyone have the latest QT development framework and know their way around C++ they should be more than capable to build and run any latest version without worry, in fact we do need any amount of help we can get, because there are actually more people asking for changes than offering help, so if you know someone who can help let them know.

**Comment by J5lx** [comment]
Maybe it would be possible to use a CI service to provide actual nightly builds. It seems that Travis is already building Pencil for Linux on every commit, so all that's left is adding builds for other operating systems and another build step for uploading them somewhere. I'm not all that familiar with packaging standalone applications, but I think it shouldn’t be too difficult to do.

**Comment by J5lx** [comment]
After having a look at the documentation of Travis it seems that they only support Linux and Mac/OS X/macOS/whatever it's called right now. So in order to have Windows builds we’d have to cross-compile the whole thing or use another CI service. I believe AppVeyor is somewhat popular for Windows builds, but I never really had a close look at it to this point.
If someone would tell me how the current nightly builds are packaged I might be able to take a look at this myself, though I can’t guarantee for anything since I don’t have a lot of time right now.

**Comment by Jose-Moreno** [comment]
@J5lx Interesting! Thank you for your offer, and don't worry about time, we're all jam packed with jobs and life. Hopefully @chchwy can also look into this to collaborate with you in this matter. Thanks again for your time and wisdom! :smile:

**Comment by CandyFace** [comment]
I have recently setup both AppVeyor and Travis-CI on one of my other projects, I'll look into this and see if I can get it working. Most of the Travis setup is there already, so it should only be a matter of getting it to build and release on github pages.

**Comment by chchwy** [comment]
Yes, we might need a server or cloud space to upload the Travis-CI builds, then we can get the linux version nightly build (which is missing atm)

**Comment by CandyFace** [comment]
Why not just use the github release page, that's what i'm currently doing for my other project. We could just make it draft if you want to control whenever they should be published or release with 0.5.4.x number automatically

**Comment by chchwy** [comment]
Oh, so there is github release page API? I am not good at these things, I would be very appreciate if you can help :P

**Comment by CandyFace** [comment]
Yeah I'm currenty looking into it, it should be as simple as adding a deploy to the script eg.
```

// travis needs to generate a tag for it to work
after_success:
  - echo -e "Starting to tag commit.\n"
  - git config --global user.email "travis@travis-ci.org"
  - git config --global user.name "Travis"
  - export GIT_TAG=build-$TRAVIS_BRANCH-$(date -u "+%Y-%m-%d-%H-%M-%S")-$TRAVIS_BUILD_NUMBER
  - git tag $GIT_TAG -a -m "Generated tag from TravisCI build $TRAVIS_BUILD_NUMBER"
  - git push origin $GIT_TAG

before_deploy:
  - cd to/my/build/path
  - zip -r "release_file.zip" Release/

deploy:
  provider: releases
  api_key: 
    secure: <insert encrypted security key>
  file: <path/to/file>
  skip_cleanup: true
  draft: false
  on:
    tags: true
    all_branches: true
    repo: Pencil2d/pencil
```

if everything goes well, then I could get this up and running today :)
That is.. if travis ever starts building, so I can test if it works before I PR :P

**Comment by J5lx** [comment]
Am I overlooking something or does that Travis script create a new *annotated* tag for *every single build*? I mean, even in the documentation of git it says “Annotated tags are meant for release […].”, and I don’t think every single build counts as actual release. Such tags would only clutter the repo. If anything, *lightweight* tags should be used for something like this IMHO, but then again I don't know whether GitHub allows marking those as releases (which can have artifacts attached to them). When it comes down to it, I’d prefer to have only actual releases stored as GitHub releases and have the last x nightly builds on some other generic file storage (Dropbox, OwnCloud, AWS, whatever) and link to them somewhere.

tl;dr: Release builds on GitHub are fine, but I think connecting every single build too tightly with the code/repo (via tags) would create clutter. Either way, any kind of nightly build is still better than none and it’s great that someone is working on it at all (thanks, @CandyFace!).

**Comment by CandyFace** [comment]
Yeah it does, it's quite annoying but that's the way Travis does it atm. I agree though that it may clutter too much. You could deploy the releases as pre-releases but then it would still clutter a bit and they would still have to be marked with tags. Travis doesn't have an api to upload to Dropbox or google drive  but you could probably make some script that does that for you. I haven't looked much into that stuff though.

**Comment by J5lx** [comment]
If it's still a thing then, I could work on an upload script on friday. I’d prefer something like Python or sh/curl though, since those languages are already present in the repo. Some maintainer (@chchwy?) would just need to tell me where the files should go (i.e. Dropbox? Google? Which directory?).

**Comment by CandyFace** [comment]
You are more than welcome. I'm done for today since I can't get Travis to get access to my git, to make a release and upload it.

**Comment by J5lx** [comment]
Umm, if I’m to work on this script…

> Some maintainer (@chchwy?) would just need to tell me where the files should go (i.e. Dropbox? Google? Which directory?).

I suppose they should go to Google Drive like the current nightlies, but I’d really like to have some confirmation on that, so I don't invest work in something that won’t be used…

**Comment by chchwy** [comment]
Thanks @J5lx, Google drive is good. Do you need me to provide anything?

**Comment by J5lx** [comment]
I got it working now, so all that’s left is setting up the API access and integrating it with CI. Especially the API access part is quite tricky though, so I’ll post a detailed guide tomorrow.

**Comment by Jose-Moreno** [comment]
@J5lx @chchwy @CandyFace Sorry to barge in like this but why Google Drive? why not use a more open platform to host the nightly builds? Also I requested to use github's release features a while ago, shouldn't that be more efficient? thoughts?

**Comment by CandyFace** [comment]
While it is possible to use Github releases, as I mentioned earlier, it can clutter the release page a lot. What other open platform would you suggest? I am not against uploading to Github Releases but as @J5lx said, maybe we should only keep those uploads for actual releases and not for each build.

The thing is, bot Travis-C and AppVeyor uses different ways to deploy their files. From what I've researched and tried already, Travis requires each upload to have its own tag and commit for each time the server builds, while AppVeyor can deploy directly to Github Releases both are quite limited in their Github integration. I haven't figured how to make them deploy onto the same release either.

Google drive is quite open already. you just goto the folder and grab the latest release.

@J5lx Nice!

**Comment by J5lx** [comment]
Mainly because that’s where nightly builds are already hosted right now. GitHub’s release feature is great for, well, releases, but it’s less than optimal for storing nightly builds (see [this comment](https://github.com/pencil2d/pencil/issues/582#issuecomment-281352583) and the next one). Most other Open Source project that provide nightly builds store them on their own server, most often via Jenkins (see e.g. [Enlightenment](https://build.enlightenment.org/), [Spigot](https://ci.md-5.net/), [NetSurf](http://ci.netsurf-browser.org/builds/)).

**Comment by Jose-Moreno** [comment]
@J5lx @CandyFace @chchwy Very well! Then it's agreed that we'll continue to use Google Drive for tthe time being. I'll research other platforms just in case, but as I said on the script commit page,  this is fantastic work, it will surely open the way for more people to test the software in a more accessible manner. Thanks a lot to all!

**Comment by J5lx** [comment]
Allright, setup guide’s ready: https://github.com/pencil2d/pencil/pull/617#issuecomment-283470017

**Comment by CandyFace** [comment]
PR for integrating @J5lx script with travis is up #627

**Comment by CandyFace** [comment]
PR up for the last missing piece, automated windows builds ٩( 'ω' )و 
I will now close this thread, as its purpose as been served.
